server/src/main/java/org/elasticsearch/index/fieldvisitor/IgnoredSourceFieldLoader.java

```diff
@@ -0,0 +1,136 @@
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.index.mapper.BlockLoader;
+import org.elasticsearch.index.mapper.IgnoredSourceFieldMapper;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+class IgnoredSourceFieldLoader extends StoredFieldLoader {
+    private final boolean forceSequentialReader;
+    private final Map<String, Set<String>> potentialFieldsInIgnoreSource;
+    private final Set<String> fieldNames;
+
+    IgnoredSourceFieldLoader(BlockLoader.FieldsSpec spec, boolean forceSequentialReader) {
+        fieldNames = new HashSet<>(spec.ignoredFieldsSpec().requiredIgnoredFields());
+        this.forceSequentialReader = forceSequentialReader;
+
+        HashMap<String, Set<String>> potentialFieldsInIgnoreSource = new HashMap<>();
+        for (String requiredIgnoredField : spec.ignoredFieldsSpec().requiredIgnoredFields()) {
+            for (String potentialStoredField : spec.ignoredFieldsSpec().format().requiredStoredFields(requiredIgnoredField)) {
+                potentialFieldsInIgnoreSource.computeIfAbsent(potentialStoredField, k -> new HashSet<>()).add(requiredIgnoredField);
+            }
+        }
+        this.potentialFieldsInIgnoreSource = potentialFieldsInIgnoreSource;
+    }
+
+    @Override
+    public LeafStoredFieldLoader getLoader(LeafReaderContext ctx, int[] docs) throws IOException {
+        var reader = forceSequentialReader ? sequentialReader(ctx) : reader(ctx, docs);
+        var visitor = new SFV(fieldNames, potentialFieldsInIgnoreSource);
+        return new LeafStoredFieldLoader() {
+
+            private int doc = -1;
+
+            @Override
+            public void advanceTo(int doc) throws IOException {
+                if (doc != this.doc) {
+                    visitor.reset();
+                    reader.accept(doc, visitor);
+                    this.doc = doc;
+                }
+            }
+
+            @Override
+            public BytesReference source() {
+                assert false : "source() is not supported by IgnoredSourceFieldLoader";
+                return null;
+            }
+
+            @Override
+            public String id() {
+                assert false : "id() is not supported by IgnoredSourceFieldLoader";
+                return null;
+            }
+
+            @Override
+            public String routing() {
+                assert false : "routing() is not supported by IgnoredSourceFieldLoader";
+                return null;
+            }
+
+            @Override
+            public Map<String, List<Object>> storedFields() {
+                return visitor.values;
+            }
+        };
+    }
+
+    @Override
+    public List<String> fieldsToLoad() {
+        return potentialFieldsInIgnoreSource.keySet().stream().toList();
+    }
+
+    static class SFV extends StoredFieldVisitor {
```

Comment:
In the future we can perhaps explore an implementation that is tuned for just fetching one stored field.